src/analysis/convex/strict_convex_space.lean

```diff
@@ -0,0 +1,128 @@
+
+See also `strict_convex_space.of_strict_convex_closed_unit_ball`. -/
+class strict_convex_space (𝕜 E : Type*) [normed_linear_ordered_field 𝕜] [normed_group E]
+  [normed_space 𝕜 E] : Prop :=
+(strict_convex_closed_ball : ∀ r : ℝ, 0 < r → strict_convex 𝕜 (closed_ball (0 : E) r))
+
+variables (𝕜 : Type*) {E : Type*} [normed_linear_ordered_field 𝕜]
+  [normed_group E] [normed_space 𝕜 E]
+
+/-- A closed ball in a strictly convex space is strictly convex. -/
+lemma strict_convex_closed_ball [strict_convex_space 𝕜 E] (x : E) (r : ℝ) :
+  strict_convex 𝕜 (closed_ball x r) :=
+begin
+  cases le_or_lt r 0 with hr hr,
+  { exact (subsingleton_closed_ball x hr).strict_convex },
+  rw ← vadd_closed_ball_zero,
+  exact (strict_convex_space.strict_convex_closed_ball r hr).vadd _,
+end
+
+variables [normed_space ℝ E]
+
+/-- A real normed vector space is strictly convex provided that the unit ball is strictly convex. -/
+lemma strict_convex_space.of_strict_convex_closed_unit_ball
+  [linear_map.compatible_smul E E 𝕜 ℝ] (h : strict_convex 𝕜 (closed_ball (0 : E) 1)) :
+  strict_convex_space 𝕜 E :=
+⟨λ r hr, by simpa only [smul_closed_unit_ball_of_nonneg hr.le] using h.smul r⟩
+
+/-- If `∥x + y∥ = ∥x∥ + ∥y∥` implies that `x y : E` are in the same ray, then `E` is a strictly
+convex space. -/
+lemma strict_convex_space.of_norm_add (h : ∀ x y : E, ∥x + y∥ = ∥x∥ + ∥y∥ → same_ray ℝ x y) :
+  strict_convex_space ℝ E :=
+begin
+  refine strict_convex_space.of_strict_convex_closed_unit_ball ℝ (λ x hx y hy hne a b ha hb hab, _),
+  have hx' := hx, have hy' := hy,
+  rw [← closure_closed_ball, closure_eq_interior_union_frontier,
+    frontier_closed_ball (0 : E) one_ne_zero] at hx hy,
+  cases hx, { exact (convex_closed_ball _ _).combo_mem_interior_left hx hy' ha hb.le hab },
+  cases hy, { exact (convex_closed_ball _ _).combo_mem_interior_right hx' hy ha.le hb hab },
+  rw [interior_closed_ball (0 : E) one_ne_zero, mem_ball_zero_iff],
+  have hx₁ : ∥x∥ = 1, from mem_sphere_zero_iff_norm.1 hx,
+  have hy₁ : ∥y∥ = 1, from mem_sphere_zero_iff_norm.1 hy,
+  have ha' : ∥a∥ = a, from real.norm_of_nonneg ha.le,
+  have hb' : ∥b∥ = b, from real.norm_of_nonneg hb.le,
+  calc ∥a • x + b • y∥ < ∥a • x∥ + ∥b • y∥ : (norm_add_le _ _).lt_of_ne (λ H, hne _)
+  ... = 1 : by simpa only [norm_smul, hx₁, hy₁, mul_one, ha', hb'],
+  have H : ∥b • y∥ • a • x = ∥a • x∥ • b • y := (h _ _ H).norm_smul_eq.symm,
+  simpa only [norm_smul, hx₁, hy₁, ha', hb', mul_one, smul_comm a, smul_right_inj ha.ne',
+    smul_right_inj hb.ne'] using H
+end
+
+variables [strict_convex_space ℝ E] {x y z : E} {a b r : ℝ}
+
+/-- If `x ≠ y` belong to the same closed ball, then a convex combination of `x` and `y` with
+positive coefficients belongs to the corresponding open ball. -/
+lemma combo_mem_ball_of_ne (hx : x ∈ closed_ball z r) (hy : y ∈ closed_ball z r) (hne : x ≠ y)
+  (ha : 0 < a) (hb : 0 < b) (hab : a + b = 1) : a • x + b • y ∈ ball z r :=
```

Comment:
why can't `hab` be ≤ 1? goes for most of these lemmas I guess; my intuition comes solely from ℝⁿ so I really wouldn't be surprised if there's some spaces where this doesn't hold

Reply:
oh, they're not centered at zero, I see

Reply:
While we can do this in `norm_combo_lt_of_ne`, I would prefer to postpone it till someone actually needs this.